Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs

```diff
@@ -0,0 +1,336 @@
+                            _npcFaction.RemoveFaction(mind.OwnedEntity.Value, "Revolutionary");
+                            _sharedStun.TryParalyze(mind.OwnedEntity.Value, stunTime, true);
+                            RemComp<RevolutionaryComponent>(mind.OwnedEntity.Value);
+                        }
+                }
+            }
+            CheckRevsWin();
+        }
+    }
+    //I LOVE REUSING CODE DIRECTLY ABOVE
+    private void CheckRevsWin()
+    {
+        var query = EntityQueryEnumerator<RevolutionaryRuleComponent, GameRuleComponent>();
+        while (query.MoveNext(out var uid, out var revs, out var gameRule))
+        {
+            if (!GameTicker.IsGameRuleAdded(uid, gameRule))
+                continue;
+
+            var heads = EntityQuery<HeadComponent, MobStateComponent>(true);
+            var aliveHeads = heads
+                .Any(ent => ent.Item2.CurrentState != MobState.Dead && ent.Item1.Running);
+            //In the rare instances that no heads are on station at start, I put a timer before this can activate. Might lower it
+            if (!aliveHeads && revs.ShuttleCalled == false && _timing.CurTime >= TimeSpan.FromMinutes(5))
+            {
+                revs.HeadsDied = true;
+                foreach (var station in _stationSystem.GetStations())
+                {
+                    _chat.DispatchStationAnnouncement(station, Loc.GetString("rev-shuttle-call"), colorOverride: Color.MediumBlue);
+                }
+                _roundEndSystem.RequestRoundEnd(null, false);
+                revs.ShuttleCalled = true;
+            }
+
+        }
+    }
+    //Should give late join command the head component. Should.
+    private void OnPlayerSpawned(PlayerSpawnCompleteEvent ev)
+    {
+        var mind = ev.Player.GetMind();
+        if (mind != null && HasComp<PendingClockInComponent>(mind.OwnedEntity) && mind.CurrentJob != null)
+        {
+            if (mind.CurrentJob.Prototype.Weight >= 10)
+            {
+                AddComp<HeadComponent>(mind.OwnedEntity.Value);
+                _sawmill.Error("Late join command added");
+            }
+        }
+    }
+
+    // Admin command
+    public void MakeHeadRev(Mind.Mind mind)
+    {
+        if (!mind.OwnedEntity.HasValue)
+            return;
+        _mindSystem.AddRole(mind, new RevolutionaryRole(mind, _prototypeManager.Index<AntagPrototype>("HeadRev")));
+        _stationSpawningSystem.EquipStartingGear(mind.OwnedEntity.Value, _prototypeManager.Index<StartingGearPrototype>("HeadRevGear"), null);
+        _npcFaction.RemoveFaction(mind.OwnedEntity.Value, "NanoTrasen");
+        _npcFaction.AddFaction(mind.OwnedEntity.Value, "Revolutionary");
+        GameTicker.AddGameRule("Revolutionary");
+        EnsureComp<RevolutionaryRuleComponent>(mind.OwnedEntity.Value);
+        EnsureComp<HeadRevolutionaryComponent>(mind.OwnedEntity.Value);
+        EnsureComp<RevolutionaryComponent>(mind.OwnedEntity.Value);
```

Comment:
See my comment about deduplication, this should be pulled into it's own function.